File: backend/routes/manager.js
import express from 'express';
import { managerAuthMiddleware } from '../auth.js';
import { run, get, all } from '../database.js';
import { logAudit } from '../business-rules.js';

const router = express.Router();

// GET /manager/dashboard - Get manager dashboard metrics
router.get('/dashboard', managerAuthMiddleware, async (req, res) => {
  try {
    const managerId = req.managerId;

    // Total Requests in System
    const totalRequestsResult = await get(
      'SELECT COUNT(*) as count FROM sales_requests'
    );
    const totalRequests = totalRequestsResult?.count || 0;

    // Pending Discount Approvals
    const pendingApprovalsResult = await get(
      `SELECT COUNT(*) as count FROM discount_requests 
       WHERE status = 'SENT_TO_MANAGER' AND requires_manager_approval = TRUE`
    );
    const pendingApprovals = pendingApprovalsResult?.count || 0;

    // Approved Today
    const approvedTodayResult = await get(
      `SELECT COUNT(*) as count FROM discount_requests 
       WHERE status = 'APPROVED' AND DATE(manager_approval_date) = CURDATE() 
       AND manager_id = ?`,
      [managerId]
    );
    const approvedToday = approvedTodayResult?.count || 0;

    // Rejected Today
    const rejectedTodayResult = await get(
      `SELECT COUNT(*) as count FROM discount_requests 
       WHERE status = 'REJECTED' AND DATE(manager_approval_date) = CURDATE()
       AND manager_id = ?`,
      [managerId]
    );
    const rejectedToday = rejectedTodayResult?.count || 0;

    // Awaiting Manager Approval (all time)
    const awaitingApprovalResult = await get(
      `SELECT COUNT(*) as count FROM discount_requests 
       WHERE requires_manager_approval = TRUE AND status = 'SENT_TO_MANAGER'`
    );
    const awaitingManagerApproval = awaitingApprovalResult?.count || 0;

    res.json({
      totalRequests,
      pendingApprovals,
      approvedToday,
      rejectedToday,
      awaitingManagerApproval
    });
  } catch (error) {
    console.error('Dashboard error:', error);
    res.status(500).json({ error: 'Failed to fetch dashboard data' });
  }
});

// GET /manager/discount-requests - Get discount requests requiring manager approval
router.get('/discount-requests', managerAuthMiddleware, async (req, res) => {
  try {
    const status = req.query.status || 'SENT_TO_MANAGER';
    let query = `
      SELECT dr.id, dr.quotation_id, dr.customer_id, dr.salesperson_id, 
             dr.requested_discount, dr.status, dr.created_at,
             c.full_name as customer_name, c.company_name,
             sp.full_name as salesperson_name,
             q.total_amount, q.quotation_status
      FROM discount_requests dr
      JOIN customers c ON dr.customer_id = c.id
      JOIN salespersons sp ON dr.salesperson_id = sp.id
      JOIN quotations q ON dr.quotation_id = q.id
      WHERE dr.requires_manager_approval = TRUE
    `;
    
    const params = [];
    
    if (status !== 'all' && status) {
      query += ` AND dr.status = ?`;
      params.push(status);
    }
    
    query += ` ORDER BY dr.created_at DESC`;
    
    const requests = await all(query, params);
    
    const formatted = requests.map(req => ({
      id: req.id,
      quotationId: req.quotation_id,
      customerId: req.customer_id,
      salespersonId: req.salesperson_id,
      customerName: req.customer_name,
      customerCompany: req.company_name,
      salespersonName: req.salesperson_name,
      requestedDiscount: req.requested_discount,
      quotationTotal: req.total_amount,
      status: req.status,
      quotationStatus: req.quotation_status,
      createdAt: new Date(req.created_at).toLocaleDateString('en-IN')
    }));

    res.json(formatted);
  } catch (error) {
    console.error('Fetch discount requests error:', error);
    res.status(500).json({ error: 'Failed to fetch discount requests' });
  }
});

// GET /manager/discount-requests/:discountRequestId - Get specific discount request details
router.get('/discount-requests/:discountRequestId', managerAuthMiddleware, async (req, res) => {
  try {
    const { discountRequestId } = req.params;

    const discountRequest = await get(
      `SELECT dr.*, 
              c.full_name as customer_name, c.company_name, c.email as customer_email,
              sp.full_name as salesperson_name, sp.email as salesperson_email,
              q.id as quotation_id, q.quotation_number, q.total_amount
       FROM discount_requests dr
       JOIN customers c ON dr.customer_id = c.id
       JOIN salespersons sp ON dr.salesperson_id = sp.id
       JOIN quotations q ON dr.quotation_id = q.id
       WHERE dr.id = ?`,
      [discountRequestId]
    );

    if (!discountRequest) {
      return res.status(404).json({ error: 'Discount request not found' });
    }

    // Get quotation line items
    const lineItems = await all(
      `SELECT * FROM quotation_line_items WHERE quotation_id = ?`,
      [discountRequest.quotation_id]
    );

    res.json({
      id: discountRequest.id,
      quotationId: discountRequest.quotation_id,
      quotationNumber: discountRequest.quotation_number,
      customerId: discountRequest.customer_id,
      salespersonId: discountRequest.salesperson_id,
      customerName: discountRequest.customer_name,
      customerCompany: discountRequest.company_name,
      customerEmail: discountRequest.customer_email,
      salespersonName: discountRequest.salesperson_name,
      salespersonEmail: discountRequest.salesperson_email,
      currentDiscount: discountRequest.current_discount_percent,
      requestedDiscount: discountRequest.requested_discount,
      quotationTotal: discountRequest.total_amount,
      reason: discountRequest.reason,
      message: discountRequest.message,
      status: discountRequest.status,
      salespersonResponse: discountRequest.salesperson_response,
      lineItems: lineItems,
      createdAt: new Date(discountRequest.created_at).toLocaleDateString('en-IN')
    });
  } catch (error) {
    console.error('Fetch discount request detail error:', error);
    res.status(500).json({ error: 'Failed to fetch discount request details' });
  }
});

// POST /manager/discount-requests/:discountRequestId/approve - Manager approves discount
router.post('/discount-requests/:discountRequestId/approve', managerAuthMiddleware, async (req, res) => {
  try {
    const { discountRequestId } = req.params;
    const { response } = req.body;
    const managerId = req.managerId;
    const userId = req.userId;

    // Get discount request
    const discountRequest = await get(
      'SELECT * FROM discount_requests WHERE id = ?',
      [discountRequestId]
    );

    if (!discountRequest) {
      return res.status(404).json({ error: 'Discount request not found' });
    }

    if (discountRequest.status !== 'SENT_TO_MANAGER') {
      return res.status(400).json({ error: 'Discount request cannot be approved in current status' });
    }

    // Update discount request
    await run(
      `UPDATE discount_requests 
       SET status = 'APPROVED', manager_id = ?, manager_response = ?, manager_approval_date = NOW()
       WHERE id = ?`,
      [managerId, response || '', discountRequestId]
    );

    // Log audit
    await logAudit(userId, 'SALES_MANAGER', 'MANAGER_APPROVED_DISCOUNT', 'discount_request', discountRequestId, {
      response,
      requestedDiscount: discountRequest.requested_discount
    });

    res.json({
      message: 'Discount approved by manager',
      status: 'APPROVED',
      discountRequestId
    });
  } catch (error) {
    console.error('Manager approve error:', error);
    res.status(500).json({ error: 'Failed to approve discount request' });
  }
});

// POST /manager/discount-requests/:discountRequestId/reject - Manager rejects discount
router.post('/discount-requests/:discountRequestId/reject', managerAuthMiddleware, async (req, res) => {
  try {
    const { discountRequestId } = req.params;
    const { response } = req.body;
    const managerId = req.managerId;
    const userId = req.userId;

    // Get discount request
    const discountRequest = await get(
      'SELECT * FROM discount_requests WHERE id = ?',
      [discountRequestId]
    );

    if (!discountRequest) {
      return res.status(404).json({ error: 'Discount request not found' });
    }

    if (discountRequest.status !== 'SENT_TO_MANAGER') {
      return res.status(400).json({ error: 'Discount request cannot be rejected in current status' });
    }

    // Update discount request
    await run(
      `UPDATE discount_requests 
       SET status = 'REJECTED', manager_id = ?, manager_response = ?, manager_approval_date = NOW()
       WHERE id = ?`,
      [managerId, response || '', discountRequestId]
    );

    // Log audit
    await logAudit(userId, 'SALES_MANAGER', 'MANAGER_REJECTED_DISCOUNT', 'discount_request', discountRequestId, {
      response,
      requestedDiscount: discountRequest.requested_discount
    });

    res.json({
      message: 'Discount rejected by manager',
      status: 'REJECTED',
      discountRequestId
    });
  } catch (error) {
    console.error('Manager reject error:', error);
    res.status(500).json({ error: 'Failed to reject discount request' });
  }
});

// POST /manager/discount-requests/:discountRequestId/counter-offer - Manager makes counter offer
router.post('/discount-requests/:discountRequestId/counter-offer', managerAuthMiddleware, async (req, res) => {
  try {
    const { discountRequestId } = req.params;
    const { counterOfferDiscount, response } = req.body;
    const managerId = req.managerId;
    const userId = req.userId;

    if (counterOfferDiscount === undefined || counterOfferDiscount < 0 || counterOfferDiscount > 100) {
      return res.status(400).json({ error: 'Invalid counter offer discount percentage' });
    }

    // Get discount request
    const discountRequest = await get(
      'SELECT * FROM discount_requests WHERE id = ?',
      [discountRequestId]
    );

    if (!discountRequest) {
      return res.status(404).json({ error: 'Discount request not found' });
    }

    if (discountRequest.status !== 'SENT_TO_MANAGER') {
      return res.status(400).json({ error: 'Discount request cannot be counter-offered in current status' });
    }

    // Update discount request with counter offer
    await run(
      `UPDATE discount_requests 
       SET status = 'COUNTER_OFFERED', manager_id = ?, counter_offer_discount = ?, 
           manager_response = ?, manager_approval_date = NOW()
       WHERE id = ?`,
      [managerId, counterOfferDiscount, response || '', discountRequestId]
    );

    // Log audit
    await logAudit(userId, 'SALES_MANAGER', 'MANAGER_COUNTER_OFFER', 'discount_request', discountRequestId, {
      counterOfferDiscount,
      originalRequest: discountRequest.requested_discount,
      response
    });

    res.json({
      message: 'Counter offer made by manager',
      status: 'COUNTER_OFFERED',
      counterOfferDiscount,
      discountRequestId
    });
  } catch (error) {
    console.error('Manager counter offer error:', error);
    res.status(500).json({ error: 'Failed to make counter offer' });
  }
});

export default router;

File: pybackend/app/routers/manager.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy import func
from decimal import Decimal
from datetime import datetime

from app.core.database import get_db
from app.dependencies.auth import get_current_manager
from app.models.discount_request import DiscountRequest
from app.models.quotation import Quotation
from app.models.sales_request import SalesRequest
from app.models.customer import Customer
from app.models.salesperson import Salesperson
from app.schemas.manager import (
    ManagerDashboardResponse, ManagerDiscountRequestResponse, 
    DetailedDiscountRequestResponse, ManagerApprovalRequest,
    ManagerRejectionRequest, ManagerCounterOfferRequest
)
from app.schemas.auth import MessageResponse

router = APIRouter(prefix="/manager", tags=["manager"])

# ==================== DASHBOARD ====================

@router.get("/dashboard", response_model=ManagerDashboardResponse)
def get_manager_dashboard(
    current: dict = Depends(get_current_manager),
    db: Session = Depends(get_db)
):
    """Get manager dashboard with approval metrics"""
    
    # Total discount requests requiring manager approval
    total_approval_requests = db.query(DiscountRequest).filter(
        DiscountRequest.requires_manager_approval == True
    ).count()
    
    # Pending approvals
    pending_approvals = db.query(DiscountRequest).filter(
        DiscountRequest.requires_manager_approval == True,
        DiscountRequest.status.in_(["Pending Review", "Requires Manager Approval"])
    ).count()
    
    # Approved today
    today = datetime.utcnow().date()
    approved_today = db.query(DiscountRequest).filter(
        DiscountRequest.manager_approval_status == "APPROVED",
        func.date(DiscountRequest.updated_at) == today
    ).count()
    
    # Rejected today
    rejected_today = db.query(DiscountRequest).filter(
        DiscountRequest.manager_approval_status == "REJECTED",
        func.date(DiscountRequest.updated_at) == today
    ).count()
    
    # Awaiting manager approvals
    awaiting_manager = db.query(DiscountRequest).filter(
        DiscountRequest.requires_manager_approval == True,
        DiscountRequest.manager_approval_status.is_(None)
    ).count()
    
    return ManagerDashboardResponse(
        totalApprovalRequests=total_approval_requests,
        pendingApprovals=pending_approvals,
        approvedToday=approved_today,
        rejectedToday=rejected_today,
        awaitingManagerApprovals=awaiting_manager
    )

# ==================== DISCOUNT REQUESTS ====================

@router.get("/discount-requests", response_model=list)
def get_manager_discount_requests(
    current: dict = Depends(get_current_manager),
    db: Session = Depends(get_db)
):
    """Get all discount requests requiring manager approval"""
    
    discount_requests = db.query(DiscountRequest).filter(
        DiscountRequest.requires_manager_approval == True
    ).all()
    
    result = []
    for dr in discount_requests:
        result.append({
            "id": dr.id,
            "quotationId": dr.quotation_id,
            "quotationNumber": None,
            "customer": None,
            "salesperson": None,
            "requestedDiscount": float(dr.requested_discount_percent),
            "currentDiscount": float(dr.current_discount_percent) if dr.current_discount_percent else 0,
            "reason": dr.reason,
            "status": dr.status,
            "createdAt": dr.created_at.isoformat() if dr.created_at else None
        })
    
    return result

@router.get("/discount-requests/{discount_request_id}", response_model=dict)
def get_manager_discount_request_detail(
    discount_request_id: str,
    current: dict = Depends(get_current_manager),
    db: Session = Depends(get_db)
):
    """Get details of a specific discount request"""
    
    discount_req = db.query(DiscountRequest).filter(
        DiscountRequest.id == discount_request_id,
        DiscountRequest.requires_manager_approval == True
    ).first()
    
    if not discount_req:
        raise HTTPException(status_code=404, detail="Discount request not found")
    
    quotation = db.query(Quotation).filter(Quotation.id == discount_req.quotation_id).first()
    customer = db.query(Customer).filter(Customer.id == discount_req.customer_id).first()
    salesperson = db.query(Salesperson).filter(Salesperson.id == discount_req.salesperson_id).first()
    
    quotation_details = None
    if quotation:
        line_items = []
        for item in quotation.line_items:
            line_items.append({
                "product": item.product_name,
                "quantity": item.quantity,
                "unitPrice": str(item.unit_price),
                "subtotal": str(item.subtotal)
            })
        
        quotation_details = {
            "quotationNumber": quotation.quotation_number,
            "total": str(quotation.total_amount),
            "lineItems": line_items
        }
    
    return {
        "id": discount_req.id,
        "quotationNumber": quotation.quotation_number if quotation else "",
        "customer": customer.full_name if customer else "",
        "customerEmail": customer.email if customer else "",
        "salesperson": salesperson.full_name if salesperson else "",
        "salespersonEmail": salesperson.email if salesperson else "",
        "currentDiscount": float(discount_req.current_discount_percent) if discount_req.current_discount_percent else 0,
        "requestedDiscount": float(discount_req.requested_discount_percent),
        "reason": discount_req.reason,
        "salespersonResponse": discount_req.salesperson_response,
        "quotationDetails": quotation_details,
        "status": discount_req.status,
        "createdAt": discount_req.created_at.isoformat() if discount_req.created_at else None
    }

@router.post("/discount-requests/{discount_request_id}/approve", response_model=MessageResponse)
def approve_discount_request(
    discount_request_id: str,
    request_data: ManagerApprovalRequest,
    current: dict = Depends(get_current_manager),
    db: Session = Depends(get_db)
):
    """Approve a discount request"""
    manager_id = current["manager_id"]
    
    discount_req = db.query(DiscountRequest).filter(
        DiscountRequest.id == discount_request_id,
        DiscountRequest.requires_manager_approval == True
    ).first()
    
    if not discount_req:
        raise HTTPException(status_code=404, detail="Discount request not found")
    
    try:
        discount_req.manager_id = manager_id
        discount_req.manager_approval_status = "APPROVED"
        discount_req.status = "Approved"
        discount_req.manager_response = request_data.response
        
        # Update quotation if needed
        quotation = db.query(Quotation).filter(Quotation.id == discount_req.quotation_id).first()
        if quotation and discount_req.requested_discount_percent > quotation.discount_percent:
            quotation.final_discount_percent = discount_req.requested_discount_percent
            quotation.quotation_status = "Awaiting Customer Response"
        
        db.commit()
        return MessageResponse(message="Discount request approved")
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to approve discount request")

@router.post("/discount-requests/{discount_request_id}/reject", response_model=MessageResponse)
def reject_discount_request(
    discount_request_id: str,
    request_data: ManagerRejectionRequest,
    current: dict = Depends(get_current_manager),
    db: Session = Depends(get_db)
):
    """Reject a discount request"""
    manager_id = current["manager_id"]
    
    discount_req = db.query(DiscountRequest).filter(
        DiscountRequest.id == discount_request_id,
        DiscountRequest.requires_manager_approval == True
    ).first()
    
    if not discount_req:
        raise HTTPException(status_code=404, detail="Discount request not found")
    
    try:
        discount_req.manager_id = manager_id
        discount_req.manager_approval_status = "REJECTED"
        discount_req.status = "Rejected"
        discount_req.manager_response = request_data.response
        db.commit()
        return MessageResponse(message="Discount request rejected")
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to reject discount request")

@router.post("/discount-requests/{discount_request_id}/counter-offer", response_model=MessageResponse)
def counter_offer_discount_request(
    discount_request_id: str,
    request_data: ManagerCounterOfferRequest,
    current: dict = Depends(get_current_manager),
    db: Session = Depends(get_db)
):
    """Submit counter-offer for a discount request"""
    manager_id = current["manager_id"]
    
    discount_req = db.query(DiscountRequest).filter(
        DiscountRequest.id == discount_request_id,
        DiscountRequest.requires_manager_approval == True
    ).first()
    
    if not discount_req:
        raise HTTPException(status_code=404, detail="Discount request not found")
    
    try:
        discount_req.manager_id = manager_id
        discount_req.counter_offer_discount_percent = Decimal(str(request_data.counterOfferDiscount))
        discount_req.manager_response = request_data.response
        discount_req.status = "Counter Offer"
        db.commit()
        return MessageResponse(message="Counter-offer submitted")
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to submit counter-offer")
